### PR TITLE
Conform to Hashable

### PR DESCRIPTION
**Pull request by aymanbagabas**
Fixes: https://github.com/Flight-School/AnyCodable/issues/45

### CODE REVIEW

File: Sources/AnyCodable/AnyCodable.swift [hunk 1]
```diff
@@ -105,3 +105,45 @@ extension AnyCodable: ExpressibleByFloatLiteral {}
...
+        case let value as Double:
+            hasher.combine(value)
+        case let value as String:
+            hasher.combine(value)
+        case let value as [String: AnyCodable]:
+            hasher.combine(value)
+        case let value as [AnyCodable]:
+            hasher.combine(value)
+        default:
+            hasher.combine(0)
```
Comment (minacle) [inline Sources/AnyCodable/AnyCodable.swift:146]: Combining `0` will result in the same hash value as when the _value_ is **really** `0`.

File: Sources/AnyCodable/AnyDecodable.swift [hunk 2]
```diff
@@ -161,3 +161,44 @@ extension AnyDecodable: CustomDebugStringConvertible {
...
+        case let value as Double:
+            hasher.combine(value)
+        case let value as String:
+            hasher.combine(value)
+        case let value as [String: AnyDecodable]:
+            hasher.combine(value)
+        case let value as [AnyDecodable]:
+            hasher.combine(value)
+        default:
+            hasher.combine(0)
```
Comment (minacle) [inline Sources/AnyCodable/AnyDecodable.swift:201]: Combining `0` will result in the same hash value as when the _value_ is **really** `0`.

File: Sources/AnyCodable/AnyEncodable.swift [hunk 3]
```diff
@@ -268,3 +268,44 @@ extension _AnyEncodable {
...
+        case let value as Double:
+            hasher.combine(value)
+        case let value as String:
+            hasher.combine(value)
+        case let value as [String: AnyEncodable]:
+            hasher.combine(value)
+        case let value as [AnyEncodable]:
+            hasher.combine(value)
+        default:
+            hasher.combine(0)
```
Comment (minacle) [inline Sources/AnyCodable/AnyEncodable.swift:308]: Combining `0` will result in the same hash value as when the _value_ is **really** `0`.

**Review by minacle — CHANGES_REQUESTED** [review]

File: Sources/AnyCodable/AnyCodable.swift (hunk 1 above)
Comment (aymanbagabas) [reply Sources/AnyCodable/AnyCodable.swift]: I've changed it to skip hashing

**Comment by minacle** [comment]
How about if the value is _Codable_ but not _Hashable_? Their hash value must not equal to others.

**Comment by 4brunu** [comment]
Looking forward to see this PR merged.
If I can somehow help, please let me know 🙂

**Review by minacle — APPROVED** [review]

**Comment by 4brunu** [comment]
Sorry, but did the CI started?
I'm asking because I see the message:
```
1 workflow awaiting approval
First-time contributors need a maintainer to approve running workflows. Learn more.
```

**Comment by minacle** [comment]
I got 500. Maybe @mattt can approve and run.

**Comment by minacle** [comment]
I ran tests per versions on my machine and every case passed. Merging.

---

## Swift 5.1.5

    Test Suite 'All tests' started at 2021-05-29 19:29:52.988
    Test Suite 'debug.xctest' started at 2021-05-29 19:29:53.035
    Test Suite 'AnyCodableTests' started at 2021-05-29 19:29:53.035
    Test Case 'AnyCodableTests.testJSONDecoding' started at 2021-05-29 19:29:53.036
    Test Case 'AnyCodableTests.testJSONDecoding' passed (0.025 seconds)
    Test Case 'AnyCodableTests.testJSONEncoding' started at 2021-05-29 19:29:53.062
    Test Case 'AnyCodableTests.testJSONEncoding' passed (0.012 seconds)
    Test Suite 'AnyCodableTests' passed at 2021-05-29 19:29:53.073
    	 Executed 2 tests, with 0 failures (0 unexpected) in 0.037 (0.037) seconds
    Test Suite 'AnyDecodableTests' started at 2021-05-29 19:29:53.074
    Test Case 'AnyDecodableTests.testJSONDecoding' started at 2021-05-29 19:29:53.074
    Test Case 'AnyDecodableTests.testJSONDecoding' passed (0.005 seconds)
    Test Suite 'AnyDecodableTests' passed at 2021-05-29 19:29:53.079
    	 Executed 1 test, with 0 failures (0 unexpected) in 0.005 (0.005) seconds
    Test Suite 'AnyEncodableTests' started at 2021-05-29 19:29:53.079
    Test Case 'AnyEncodableTests.testEncodeNSNumber' started at 2021-05-29 19:29:53.079
    Test Case 'AnyEncodableTests.testEncodeNSNumber' passed (0.004 seconds)
    Test Case 'AnyEncodableTests.testJSONEncoding' started at 2021-05-29 19:29:53.084
    Test Case 'AnyEncodableTests.testJSONEncoding' passed (0.004 seconds)
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' started at 2021-05-29 19:29:53.087
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' passed (0.003 seconds)
    Test Suite 'AnyEncodableTests' passed at 2021-05-29 19:29:53.090
    	 Executed 3 tests, with 0 failures (0 unexpected) in 0.01 (0.01) seconds
    Test Suite 'debug.xctest' passed at 2021-05-29 19:29:53.090
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.052 (0.052) seconds
    Test Suite 'All tests' passed at 2021-05-29 19:29:53.090
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.052 (0.052) seconds

## Swift 5.2.5

    Test Suite 'All tests' started at 2021-05-29 19:35:06.082
    Test Suite 'debug.xctest' started at 2021-05-29 19:35:06.128
    Test Suite 'AnyCodableTests' started at 2021-05-29 19:35:06.128
    Test Case 'AnyCodableTests.testJSONDecoding' started at 2021-05-29 19:35:06.128
    Test Case 'AnyCodableTests.testJSONDecoding' passed (0.024 seconds)
    Test Case 'AnyCodableTests.testJSONEncoding' started at 2021-05-29 19:35:06.153
    Test Case 'AnyCodableTests.testJSONEncoding' passed (0.011 seconds)
    Test Suite 'AnyCodableTests' passed at 2021-05-29 19:35:06.165
    	 Executed 2 tests, with 0 failures (0 unexpected) in 0.035 (0.035) seconds
    Test Suite 'AnyDecodableTests' started at 2021-05-29 19:35:06.165
    Test Case 'AnyDecodableTests.testJSONDecoding' started at 2021-05-29 19:35:06.165
    Test Case 'AnyDecodableTests.testJSONDecoding' passed (0.004 seconds)
    Test Suite 'AnyDecodableTests' passed at 2021-05-29 19:35:06.170
    	 Executed 1 test, with 0 failures (0 unexpected) in 0.004 (0.004) seconds
    Test Suite 'AnyEncodableTests' started at 2021-05-29 19:35:06.170
    Test Case 'AnyEncodableTests.testEncodeNSNumber' started at 2021-05-29 19:35:06.170
    Test Case 'AnyEncodableTests.testEncodeNSNumber' passed (0.003 seconds)
    Test Case 'AnyEncodableTests.testJSONEncoding' started at 2021-05-29 19:35:06.173
    Test Case 'AnyEncodableTests.testJSONEncoding' passed (0.003 seconds)
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' started at 2021-05-29 19:35:06.176
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' passed (0.002 seconds)
    Test Suite 'AnyEncodableTests' passed at 2021-05-29 19:35:06.178
    	 Executed 3 tests, with 0 failures (0 unexpected) in 0.008 (0.008) seconds
    Test Suite 'debug.xctest' passed at 2021-05-29 19:35:06.178
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds
    Test Suite 'All tests' passed at 2021-05-29 19:35:06.178
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds

## Swift 5.3.3

    Test Suite 'All tests' started at 2021-05-29 19:39:25.727
    Test Suite 'debug.xctest' started at 2021-05-29 19:39:25.764
    Test Suite 'AnyCodableTests' started at 2021-05-29 19:39:25.765
    Test Case 'AnyCodableTests.testJSONDecoding' started at 2021-05-29 19:39:25.765
    Test Case 'AnyCodableTests.testJSONDecoding' passed (0.022 seconds)
    Test Case 'AnyCodableTests.testJSONEncoding' started at 2021-05-29 19:39:25.788
    Test Case 'AnyCodableTests.testJSONEncoding' passed (0.011 seconds)
    Test Suite 'AnyCodableTests' passed at 2021-05-29 19:39:25.799
    	 Executed 2 tests, with 0 failures (0 unexpected) in 0.033 (0.033) seconds
    Test Suite 'AnyDecodableTests' started at 2021-05-29 19:39:25.800
    Test Case 'AnyDecodableTests.testJSONDecoding' started at 2021-05-29 19:39:25.800
    Test Case 'AnyDecodableTests.testJSONDecoding' passed (0.005 seconds)
    Test Suite 'AnyDecodableTests' passed at 2021-05-29 19:39:25.805
    	 Executed 1 test, with 0 failures (0 unexpected) in 0.005 (0.005) seconds
    Test Suite 'AnyEncodableTests' started at 2021-05-29 19:39:25.805
    Test Case 'AnyEncodableTests.testEncodeNSNumber' started at 2021-05-29 19:39:25.805
    Test Case 'AnyEncodableTests.testEncodeNSNumber' passed (0.004 seconds)
    Test Case 'AnyEncodableTests.testJSONEncoding' started at 2021-05-29 19:39:25.809
    Test Case 'AnyEncodableTests.testJSONEncoding' passed (0.004 seconds)
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' started at 2021-05-29 19:39:25.812
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' passed (0.003 seconds)
    Test Suite 'AnyEncodableTests' passed at 2021-05-29 19:39:25.815
    	 Executed 3 tests, with 0 failures (0 unexpected) in 0.01 (0.01) seconds
    Test Suite 'debug.xctest' passed at 2021-05-29 19:39:25.815
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds
    Test Suite 'All tests' passed at 2021-05-29 19:39:25.815
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds

## Swift 5.4.1

    Test Suite 'All tests' started at 2021-05-29 19:42:18.164
    Test Suite 'debug.xctest' started at 2021-05-29 19:42:18.201
    Test Suite 'AnyCodableTests' started at 2021-05-29 19:42:18.202
    Test Case 'AnyCodableTests.testJSONDecoding' started at 2021-05-29 19:42:18.202
    Test Case 'AnyCodableTests.testJSONDecoding' passed (0.024 seconds)
    Test Case 'AnyCodableTests.testJSONEncoding' started at 2021-05-29 19:42:18.227
    Test Case 'AnyCodableTests.testJSONEncoding' passed (0.012 seconds)
    Test Suite 'AnyCodableTests' passed at 2021-05-29 19:42:18.239
    	 Executed 2 tests, with 0 failures (0 unexpected) in 0.036 (0.036) seconds
    Test Suite 'AnyDecodableTests' started at 2021-05-29 19:42:18.240
    Test Case 'AnyDecodableTests.testJSONDecoding' started at 2021-05-29 19:42:18.240
    Test Case 'AnyDecodableTests.testJSONDecoding' passed (0.003 seconds)
    Test Suite 'AnyDecodableTests' passed at 2021-05-29 19:42:18.243
    	 Executed 1 test, with 0 failures (0 unexpected) in 0.003 (0.003) seconds
    Test Suite 'AnyEncodableTests' started at 2021-05-29 19:42:18.243
    Test Case 'AnyEncodableTests.testEncodeNSNumber' started at 2021-05-29 19:42:18.243
    Test Case 'AnyEncodableTests.testEncodeNSNumber' passed (0.003 seconds)
    Test Case 'AnyEncodableTests.testJSONEncoding' started at 2021-05-29 19:42:18.246
    Test Case 'AnyEncodableTests.testJSONEncoding' passed (0.003 seconds)
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' started at 2021-05-29 19:42:18.249
    Test Case 'AnyEncodableTests.testStringInterpolationEncoding' passed (0.002 seconds)
    Test Suite 'AnyEncodableTests' passed at 2021-05-29 19:42:18.252
    	 Executed 3 tests, with 0 failures (0 unexpected) in 0.008 (0.008) seconds
    Test Suite 'debug.xctest' passed at 2021-05-29 19:42:18.252
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds
    Test Suite 'All tests' passed at 2021-05-29 19:42:18.252
    	 Executed 6 tests, with 0 failures (0 unexpected) in 0.048 (0.048) seconds

**Pull request merged**